src/UI/Implementation/Render/FSLoader.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/* Copyright (c) 2017 Richard Klees <richard.klees@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Implementation\Render;
+
+use ILIAS\UI\Factory as RootFactory;
+use ILIAS\UI\Component\Component;
+
+/**
+ * Loads renderers for components from the file system.
+ */
+class FSLoader implements Loader {
+	use LoaderHelper;
+
+	/**
+	 * @var	RootFactory
+	 */
+	private $ui_factory;
+
+	/**
+	 * @var	Render\TemplateFactory
+	 */
+	private $tpl_factory;
+
+	/**
+	 * @var	\ilLanguage
+	 */
+	private $lng;
+
+	/**
+	 * @var	Render\JavaScriptBinding
+	 */
+	private $js_binding;
+
+	public function __construct(RootFactory $ui_factory, TemplateFactory $tpl_factory, \ilLanguage $lng, JavaScriptBinding $js_binding) {
+        $this->ui_factory = $ui_factory;
+        $this->tpl_factory = $tpl_factory;
+        $this->lng = $lng;
+        $this->js_binding = $js_binding;
+    }
+
+	/**
+	 * @inheritdocs
+	 */
+	public function getRendererFor(Component $component, array $contexts) {
+		$class = get_class($component);
+		$context_names = $this->getContextNames($contexts);
+		$renderer_classes = $this->getRendererNamesFor($class, $context_names);
+		foreach ($renderer_classes as $renderer_class) {
+			if (class_exists($renderer_class)) {
+				return new $renderer_class
+					($this->ui_factory
+					, $this->tpl_factory
+					, $this->lng
+					, $this->js_binding
+					);
+			}
+		}
+		throw new \LogicException("No rendered for '".$class."' found.");
+    }
+
+	/**
+	 * Get the possible class names for the renderer of Component class under the given
+	 * contexts.
+	 *
+	 * @param	string		$class
+	 * @param	string[]	$contexts
+	 * @return	string
+	 */
+	protected function getRendererNamesFor($class, array $contexts) {
+		$parts = explode("\\", $class);
+		$parts[count($parts)-1] = "Renderer";
+		$base = implode("\\", $parts);
+		if (count($contexts) == 0) {
+			return [$base];
+		}
+		$ret = [$base."_".implode("_", $contexts)];
+		$last = array_pop($contexts);
+		while($last) {
+			$ret[] = $base."_".$last;
+			$last = array_pop($contexts);
+		}
+		$ret[] = $base;
+		return $ret;
+	}
```

Comment:
By ["ContextA", "ContextB", "ContextC"] you mean, that this is the order in which they were added, i.e. ContextA was added first, right? Then your understanding is correct. The general idea was to go from very specific ("Base_ContextA_ContextB_ContextC") to less specific ("Base"). You are definitely right about the possibility that a very specific context would stop working if another context is added somewhere in the middle.

The reason to chose the current combinations was threefold:
* Every additional combination potentially causes an access to the file system at the moment, especially for the default case, where I have no specific renderer. Specific renderers need to be checked for existence before more general once. Thus I wanted to have only small amount of possible renderer names in general.
* With your order it seems obvious to me, that a component is in ContextC more than in ContextB, so ContextC is checked first.
* For the use case we currently know (Listings with Items), one level of nested context seems to be enough. Another use case for this feature most possibly could be a kiosk-mode, this is were two contexts will be nested regarding Listings and Items. For this cases I wanted to maintain the possibility to override the Listings-Renderer for the Kiosk-Mode. Thus the first very specific name.

So, yes, there is possible problem hidden here, but tradeoffs. Do you have any other ideas, how renderer names could be chosen? Do we want other tradeoffs? Any opinions from @Amstutz about this?

Best regards!